drivers/infiniband/hw/hfi1/mmu_rb.c

```diff
@@ -157,25 +157,23 @@ void hfi1_mmu_rb_unregister(struct mmu_rb_handler *handler)
 	struct list_head del_list;
 
 	/* Unregister first so we don't get any more notifications. */
+	/*
+	 * Must ensure that the mm is not freed before do_remove(),
+	 * so we keep one mmgrab here and do the mmdrop later.
+	 * Except for the !NO_NOTIFIER_REG_GRAB case, we already
+	 * have an extra mmgrab at this point.
+	 */
 	if (handler->registered) {
 #ifdef NO_MMU_NOTIFIER_MM
-		mmu_notifier_unregister(&handler->mn, handler->mm);
-#else
-		mmu_notifier_unregister(&handler->mn, handler->mn.mm);
+#ifndef NO_NOTIFIER_REG_GRAB
```

Comment:
I think this basically needs to be the same as the drop logic. We always want to do a grab/ungrab. That's the fix right? So the only difference between distros is if the mm is in the notifier or directly in our handler.

Reply:
We don't NEED to do the mmgrab unless the only reference we have was by mmu_notifier_register. So, we could unconditionally do an mmgrab before any of the unregister stuff, and leave the original conditional mmdrop. The resulting code would be a little odd/sub-optimal in that it does mmgrab, mmdrop, mmdrop. But, might be clearer as far as diffs are concerned. Would you rather do it that way then?

Reply:
So NO_NOTIFER_REG_GRAB means we did an explicit mmgrab in our code. So if that's NOT on then we have to do the grab here. Ok makes sense.